deploy/tests/test_yaml_api.py

```diff
@@ -0,0 +1,402 @@
+
+    def __init__(self, hwdb, run):
+        self.hwdb = hwdb
+        self.run = run
+
+    def write(self):
+        self.hwdb.dump()
+        self.run.dump()
+
+    @property
+    def args(self):
+        return ['-a', fspath(self.hwdb.path),
+                '-c', fspath(self.run.path)]
+
+    @property
+    def cmdline(self):
+        return shjoin(self.args)
+
+@pytest.fixture()
+def sample_backup_configs() -> Path:
+    dir = Path(__file__).parent.parent / 'sample-backup-configs'
+    dir.is_dir().should.equal(True)
+    return dir
+
+@pytest.fixture()
+def scy_build(tmp_path: Path, sample_backup_configs: Path) -> TmpYaml:
+    return TmpYaml(tmp_path, sample_backup_configs / 'sample_config_build.yaml')
+
+@pytest.fixture()
+def scy_build_recipes(tmp_path: Path, sample_backup_configs: Path) -> TmpYaml:
+    return TmpYaml(tmp_path, sample_backup_configs / 'sample_config_build_recipes.yaml')
+
+@pytest.fixture()
+def scy_build_farm(tmp_path: Path, sample_backup_configs: Path) -> TmpYaml:
+    return TmpYaml(tmp_path, sample_backup_configs / 'sample_config_build_farm.yaml')
+
+@pytest.fixture()
+def build_yamls(scy_build, scy_build_recipes, scy_build_farm, scy_hwdb) -> BuildTmpYamlSet:
+    return BuildTmpYamlSet(scy_build, scy_build_recipes, scy_build_farm, scy_hwdb)
+
+@pytest.fixture()
+def scy_hwdb(tmp_path: Path, sample_backup_configs: Path) -> TmpYaml:
+    return TmpYaml(tmp_path, sample_backup_configs / 'sample_config_hwdb.yaml')
+
+@pytest.fixture()
+def scy_runtime(tmp_path: Path, sample_backup_configs: Path) -> TmpYaml:
+    return TmpYaml(tmp_path, sample_backup_configs / 'sample_config_runtime.yaml')
+
+@pytest.fixture()
+def run_yamls(scy_hwdb: TmpYaml, scy_runtime: TmpYaml) -> RunTmpYamlSet:
+    return RunTmpYamlSet(scy_hwdb, scy_runtime)
+
+@pytest.fixture()
+def non_existent_file(tmp_path):
+    # tmp_path is builtin pytest fixture to get a per-test temporary directory that should be clean
+    # but we still make sure that it doesn't exist before giving it
+    file = tmp_path / 'GHOST_FILE'
+    file.exists().should.equal(False)
+    return file
+
+@pytest.fixture()
+def firesim_argparser():
+    return firesim.construct_firesim_argparser()
+
+@pytest.fixture()
+def fs_cl2args(firesim_argparser):
+    """fixture that bundles calling parse_args and shlex.split to abbreviate usage in tests"""
+    return lambda x: firesim_argparser.parse_args(shlex.split(x))
+
+@pytest.mark.usefixtures("aws_test_credentials")
+class TestConfigBuildAPI:
+    """ Test config_{build, build_recipes}.yaml APIs """
+
+    # run_test("invalid-build-section")
+    @patch('firesim.buildafi')
+    def test_invalid_build_section(self, buildafi_mock, build_yamls, fs_cl2args):
+        # @patch modifies the symbol table with our Mock but we're dispatching through
+        # a dict and that needs to be patched as well, otherwise our mock won't be called but
+        # the real function will be
+        firesim.BUILD_TASKS['buildafi'] = buildafi_mock
```

Comment:
What defines this `buildafi_mock` (fixture right)? This is just removing the calls to the AWS EC2 API right?

Reply:
The `@patch` decorator passes it in as the first fixture.  I kinda dislike writing it this way and I'm remembering that I dislike it because it's not obvious that `@patch` is doing that.  I'll look at how to write this a better way.

Reply:
Ok.  I went back and reviewed some things.  I think the pytest-mock plugin makes this code look much better.  I was going to use it when I added the pytests but was hesitant to add yet another dependency.  However, looking back at https://github.com/pytest-dev/pytest-mock/#why-bother-with-a-plugin I'm really seeing how much nicer it would be.  I'm going to add that quick this afternoon and finish implementing the rest of the tests so we can merge this.